### PR TITLE
Update readout example notebook

### PR DESCRIPTION
**Pull request by joshcombes**

### CODE REVIEW

**Comment by joshcombes** [comment]
will partially close https://github.com/rigetti/forest-benchmarking/issues/98

File: examples/readout_fidelity.ipynb [hunk 1]
```diff
@@ -4,21 +4,116 @@
...
+    "0 &  1\n",
+    "\\end{pmatrix}.\n",
+    "$$\n",
+    "Imperfect measurements have $ \\Pr( j | i ) = p_{j,i} $  where $\\sum_j  \\Pr( j | i ) = 1$ for any $i$.  \n",
+    "  \n",
+    "  \n",
+    "**In summary the functionality of `readout.py` is:**\n",
+    "- estimation of a single qubit confusion matrix\n",
+    "- estimation of the joint confusion matrix over `n` bits\n",
+    "- marginalize a joint confusion matrix on `n` to `k` confusion matrices over `m` qubits\n",
```
Comment (kylegulshen) [inline examples/readout_fidelity.ipynb]:
```suggestion
    "- marginalize a joint confusion matrix over `m` qubits to a smaller matrix over a subset of those qubits\n",
```

File: examples/readout_fidelity.ipynb [hunk 2]
```diff
@@ -4,21 +4,116 @@
...
+    "$$\n",
+    "C = \\begin{pmatrix}\n",
+    "p(0 | 0)   & p(0 | 1)   & \\cdots & p(0 | {d-1})  \\\\\n",
+    "p(1 | 0)   & p(1 | 1)   & \\cdots & p(1 | {d-1})  \\\\\n",
+    "\\vdots       &              &        & \\vdots          \\\\\n",
+    "p(d-1 | 0) & p(d-1 | 1) & \\cdots & p(d-1 | {d-1})\n",
+    "\\end{pmatrix}.\n",
+    "$$\n",
+    "\n",
+    "The trace of the confusion rate matrix divided by the number of states\n",
```
Comment (kylegulshen) [inline examples/readout_fidelity.ipynb]:
```suggestion
    "The trace of the confusion matrix divided by the number of states\n",
```

File: examples/readout_fidelity.ipynb [hunk 3]
```diff
@@ -4,21 +4,116 @@
...
+    "is the average probability to correctly report the input string\n",
+    "\n",
+    "$$\n",
+    "F(C):=\\frac{\\mathrm{Tr}\\left( C\\right)}{d} = \\frac{1}{d} \\sum_{j=0}^{d-1} p(j|j)\n",
+    "$$\n",
+    "this is some times called the \"joint assignment fidelity\" or \"readout fidelity\" of our simultaneous qubit readout.\n",
+    "\n",
+    "This matrix can be related to the quantum measurement operators (well the [POVM](https://en.wikipedia.org/wiki/POVM))\n",
+    "Given the coefficients appearing in the confusion rate matrix the equivalent readout POVM is\n",
```
Comment (kylegulshen) [inline examples/readout_fidelity.ipynb]:
```suggestion
    "Given the coefficients appearing in the confusion matrix the equivalent readout POVM is\n",
```

File: examples/readout_fidelity.ipynb [hunk 4]
```diff
@@ -4,21 +4,116 @@
...
+    "\n",
+    "The trace of the confusion rate matrix divided by the number of states\n",
+    "is the average probability to correctly report the input string\n",
+    "\n",
+    "$$\n",
+    "F(C):=\\frac{\\mathrm{Tr}\\left( C\\right)}{d} = \\frac{1}{d} \\sum_{j=0}^{d-1} p(j|j)\n",
+    "$$\n",
+    "this is some times called the \"joint assignment fidelity\" or \"readout fidelity\" of our simultaneous qubit readout.\n",
+    "\n",
+    "This matrix can be related to the quantum measurement operators (well the [POVM](https://en.wikipedia.org/wiki/POVM))\n",
```
Comment (kylegulshen) [inline examples/readout_fidelity.ipynb]:
```suggestion
    "This matrix can be related to the quantum measurement operators (well the [POVM](https://en.wikipedia.org/wiki/POVM)).\n",
```

File: examples/readout_fidelity.ipynb [hunk 5]
```diff
@@ -36,14 +131,12 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## measuring readout fidelity\n",
-    "\n",
     "`run_readout` constructs and measures a program for a single qubit depending on whether the $\\left|0\\right>$ state or the $\\left|1\\right>$ state is targeted. When $\\left|0\\right>$ is targeted, we prepare a qubit in the ground state with an identity operation. Similarly, When $\\left|1\\right>$ is targeted, we prepare a qubit in the excited state with an `X` gate. "
```
Comment (kylegulshen) [inline examples/readout_fidelity.ipynb]:
```suggestion
    "`run_readout` constructs and measures a program for a single qubit depending on whether the $| 0 \\rangle$ state or the $|1 \\rangle$ state is targeted. When $|0 \\rangle$ is targeted, we prepare a qubit in the ground state with an identity operation. Similarly, when $|1\\rangle$ is targeted, we prepare a qubit in the excited state with an `X` gate. "
```

File: examples/readout_fidelity.ipynb [hunk 6]
```diff
@@ -90,7 +183,7 @@
   },
```
Comment (kylegulshen) [inline examples/readout_fidelity.ipynb]: Line 181 needs to be corrected as well. Latex + 'targetting'

**Comment by joshcombes** [comment]
@kylegulshen you might want to look at the changes I just made.

File: examples/readout_fidelity.ipynb [hunk 7]
```diff
@@ -4,21 +4,116 @@
...
+    "# Readout Fidelity Measurement \n",
     "\n",
-    "Run an identity, `I`, or an `X` pulse to estimate readout fidelity. "
+    "In this notebook we explore the module `readout.py` that enable \n",
+    "easy estimation of readout imperfections in the computational basis.\n",
+    "\n",
+    "The basic idea is to prepare all possible computational basis states \n",
+    "and measure the resulting distribution over output strings. For \n",
+    "input bit strings $i$ and output strings $j$ we want learn all \n",
+    "values of $\\Pr( {\\rm detected\\ } j\n",
```
[the remark this answers is not in the document]
Comment (joshcombes) [reply examples/readout_fidelity.ipynb]: maybe the most recent change I pushed will fix it.

File: examples/readout_fidelity.ipynb [hunk 8]
```diff
@@ -4,149 +4,179 @@
...
+    "where we have introduced the bitstring projectors\n",
+    "$ \\Pi_{k}=|k\\rangle \\langle k|$. \n",
+    "\n",
+    "\n",
+    "To be a valid POVM we must have \n",
+    "1. $\\hat N_j\\ge 0$ for all $j$, and \n",
+    "2. $\\sum_j\\hat N_j\\le I$, \n",
+    "\n",
+    "where $I$ is the identity operator. We can immediately see \n",
+    "condition one is true. While condition two is easy to verify\n",
```
Comment (kylegulshen) [inline examples/readout_fidelity.ipynb]:
```suggestion
    "condition one is true. Condition two is easy to verify:\n",
```

File: examples/readout_fidelity.ipynb [hunk 9]
```diff
@@ -4,149 +4,179 @@
...
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "## measuring readout fidelity\n",
+    "## Simple readout fidelity and confusion matrix estimation\n",
     "\n",
-    "`run_readout` constructs and measures a program for a single qubit depending on whether the $\\left|0\\right>$ state or the $\\left|1\\right>$ state is targeted. When $\\left|0\\right>$ is targeted, we prepare a qubit in the ground state with an identity operation. Similarly, When $\\left|1\\right>$ is targeted, we prepare a qubit in the excited state with an `X` gate. "
+    "The simple example below should help understanding the more general code in `readout.py`."
```
Comment (kylegulshen) [inline examples/readout_fidelity.ipynb]:
```suggestion
    "The simple example below should help with understanding the more general code in `readout.py`."
```

File: examples/readout_fidelity.ipynb [hunk 10]
```diff
@@ -169,68 +199,271 @@
...
+    "    [[0.925, 0.025, 0.025, 0.025],\n",
+    "     [0.000, 1.000, 0.000, 0.000],\n",
+    "     [0.000, 0.000, 1.000, 0.000],\n",
+    "     [0.000, 0.000, 0.000, 1.000]])\n",
+    "\n",
+    "\n",
+    "marginal1 = marginalize_confusion_matrix(confusion_correlated_two_q_ro_errors, [0, 1], [0])\n",
+    "marginal2 = marginalize_confusion_matrix(confusion_correlated_two_q_ro_errors, [0, 1], [1])\n",
+    "\n",
+    "# take the two maringal confusion matricies and recombine\n",
```
Comment (kylegulshen) [inline examples/readout_fidelity.ipynb]:
```suggestion
    "# take the two marginal confusion matrices and recombine\n",
```

**Pull request merged**